### PR TITLE
Upload delay

### LINKED ISSUE #76: MiSeq Analysis finishes AFTER it indicates the run is complete
## Describe the bug
When the MiSeq finishes analysis of a run, we believed it writes an .xml file that indicates it is finished, but this is not the case. We have found it can take up to 20 minutes after indicating complete to actually complete.

## Solution
We have noticed this can become an issue when automating uploads, and I have built a temporary fix.
https://github.com/phac-nml/irida-uploader/tree/timeout_batch_runs
This lets you use `--timeout 60` to delay runs that are discovered by an hour, to ensure that runs have the time needed to complete.

In the future we will need a more concrete solution that KNOWS when a run is finished, not just guessing that it will take under an hour after it notifies it's complete.

### PR DESCRIPTION
**Pull request by JeffreyThiessen**
## Description of changes
Added another directory status `delayed` and a new config option `delay = {INTEGER}`

When a `delay` > 0 is set, runs that are found as new will be set to `delayed`.
When a `delayed` run is found by the uploader, it will check if `delay` minutes have passed before uploading

Improved the status file to include per-sample upload information

the `DirectoryStatus` object, `upload_status.py`, and `cli_entry.py` files were refactored to support these changes

## Related issue
Fixes #76 

## Checklist
Things for the developer to confirm they've done before the PR should be accepted:

* [x] CHANGELOG.md updated with information for new change.
* [x] Tests added (or description of how to test) for any new features.
* [x] User documentation updated for UI or technical changes.

### CODE REVIEW

File: scripts/update_version.sh [hunk 1]
```diff
@@ -12,10 +12,8 @@ setuppy=setup.py
 sed -i "s/version=.*\,/version='$newversion',/" "$setuppy"
 
 wininstall=windows-installer.cfg
-wininstallgui=windows-gui-installer.cfg
```
Comment (JeffreyThiessen) [inline scripts/update_version.sh:15]: the cli and gui windows installers were combined into a single installer in 0.5.0, this script was not updated to reflect that by mistake.

File: setup.py [hunk 2]
```diff
@@ -28,7 +28,7 @@
                       ],
     extras_require={
         "GUI": ["PyQt5==5.12.1", "PyQt5-stubs==5.12.1"],
-        "TEST": ["selenium"],
+        "TEST": ["selenium", "pytest"],
```
Comment (JeffreyThiessen) [inline setup.py:31]: Technically, this is not needed for tests to run, but it does allow for a developer to run a single test manually after a suite fails so it's a nice QOL improvement.

File: iridauploader/tests_integration/integration_data_setup.py [hunk 3]
```diff
@@ -155,7 +155,7 @@ def create_client(self):
 
     def get_irida_secret(self):
 
-        self.driver.get(self.base_url + "/clients")
+        self.driver.get(self.base_url + "/admin/clients")
```
Comment (JeffreyThiessen) [inline iridauploader/tests_integration/integration_data_setup.py:158]: New IRIDA version moved where the clients page sits.

File: iridauploader/core/upload.py [hunk 4]
```diff
@@ -0,0 +1,253 @@
+"""
```
Comment (JeffreyThiessen) [inline iridauploader/core/upload.py:1]: github is not showing it, but this file was cli_entry.py, renamed and heavily edited.

File: iridauploader/progress/upload_status.py [hunk 5]
```diff
@@ -14,114 +13,93 @@
...
     """
-    result = DirectoryStatus(directory)
-
     # Verify directory is readable
     if not os.access(directory, os.R_OK):
-        result.status = DirectoryStatus.INVALID
-        result.message = 'Directory cannot be read. Please check permissions'
-        return result
+        directory_status = DirectoryStatus(directory)
+        directory_status.status = DirectoryStatus.INVALID
```
Comment (tom114) [inline iridauploader/progress/upload_status.py:31]: I'm seeing pretty much every time you instantiate a `DirectoryStatus` a `directory_status.status = DirectoryStatus.WHATEVER` right after.  Do we want to have a status required in the constructor here?  If it's going to be set every time it's instantiated it likely should be a constructor param.

File: iridauploader/core/upload.py [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+        logging.error(error_msg)
+        return exit_error(error_msg)
+    # Check if run is new, if delay config is > 0, delay the run, exit with success
+    elif directory_status.status_equals(DirectoryStatus.NEW):
+        if delay_minutes > 0:
+            upload_helpers.set_run_delayed(directory_status)
+            logging.info("Run has been delayed for {} minutes.".format(delay_minutes))
+            return exit_success()
+    # If run was delayed, check if run can now be uploaded, if not, exit with success
+    elif directory_status.status_equals(DirectoryStatus.DELAYED):
```
Comment (tom114) [inline iridauploader/core/upload.py:53]: I feel like this big conditional has made this method and `batch_upload_single_entry` overly messy.  They used to be pretty small tight methods for the tasks they're doing.  I think this should be refactored somehow,

Does it make sense to take that code out into another component or something?  Maybe move it out into the `upload_helpers` module or something so this method can just ask "is this run ready to go?"

File: iridauploader/core/upload.py [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+    logging.info("Uploads completed with {} error(s)".format(len(error_list)))
+    for directory_status in error_list:
+        logging.warning("Directory '{}' upload exited with ERROR, check log and status file for details"
+                        "".format(directory_status.directory))
+
+    logging.info("Batch upload complete, Exiting!")
+    return exit_success()
+
+
+def _validate_and_upload(directory_status, upload_mode):
```
Comment (tom114) [inline iridauploader/core/upload.py:165]: Big fan of how much more concise this method is.  Thanks for the refactor here.

**Review by tom114 — CHANGES_REQUESTED** [review]
Couple issues and questions.  Thanks, functionally the delay seems to work as intended!

File: iridauploader/gui/threads.py [hunk 8]
```diff
@@ -139,9 +140,19 @@ def set_vars(self, run_dir, force_state, upload_mode):
...
         This runs when the threads start call is done
+
+        Also temporarily sets delay to 0 while uploading to avoid strange interactions.
         :return:
         """
-        self._exit_return = cli_entry.upload_run_single_entry(self._run_dir, self._force_state, self._upload_mode)
+        # Store the current delay config setting
+        delay_config_int = int(config.read_config_option("delay", 0))
+        # Temporarily set delay to 0, s.t. upload will proceed even if command line is configured to delay runs
+        config.set_config_options(delay=0)
```
[the remark this answers is not in the document]
Comment (JeffreyThiessen) [reply iridauploader/gui/threads.py:150]: So the issue is that we don't want to overwrite the config file's delay value for when doing automated uploads and command line uploads, but we want the gui uploader to be able to upload as soon as the user clicks upload.

The functions in upload.py are using the config module to determine if a run should be delayed, or proceed with upload.
So the solution is to change the delay value to 0 in the config module, run the upload and then set it back to the value it was.

We set it temporarily s.t. if the user changes settings later on, the config module option doesn't get written to the config file proper. 

I'm not 100% sold on this functionality flow either, but the only alternative I could think of would be to add another parameter to the single entry upload that added more logic to skip the delay checks/steps, which I was less a fan of.

File: iridauploader/core/upload.py (hunk 7 above)
Comment (JeffreyThiessen) [reply iridauploader/core/upload.py:165]: No problem, I wish I had taken a picture of my whiteboard while I was figuring out how to redo the code flow and exception handling for it.

File: iridauploader/progress/upload_status.py (hunk 5 above)
Comment (JeffreyThiessen) [reply iridauploader/progress/upload_status.py]: Good call, I added both `status` and `message` params to the constructor

File: iridauploader/core/cli.py [hunk 9]
```diff
@@ -93,6 +99,13 @@ def init_argparser():
     argument_parser.add_argument('-r', '--readonly',
                                  action='store_true',  # This line makes it not parse a variable
                                  help='Upload in Read Only mode, does not write status or log file to run directory.')
+    argument_parser.add_argument('-cd', '--delay',
```
[the remark this answers is not in the document]
Comment (JeffreyThiessen) [reply iridauploader/core/cli.py]: Fixed it. Now it behaves like the other `-c*` options

[the remark this answers is not in the document]
Comment (tom114) [reply iridauploader/core/cli.py]: Works now.  Thanks.

**Review by tom114 — APPROVED** [review]

**Pull request merged**